README
$Id$

ChangeLog
----------------------------
Changelog file

README
----------------------------
This file

singsing.c
----------------------------
singsing core engine

singsing.h
----------------------------
singsing public include

singsing_p.h
----------------------------
singsing private include

examples/filippo.c
----------------------------
A simple socks v4&v5 scanner

examples/ssh_scan.c
----------------------------
SSH scanner (no banner grabbing)

examples/zucca.c
----------------------------
syn scanner

examples/zuccorale.c
----------------------------
oracle scanner

examples/telnet_grabber.c
----------------------------
telnet banner grabbing tool


ChangeLog
$Id$

Version 0.5:
------------------------
* Added zuccorale.c on example
* Added time statistics on zucca
* Removed all global variables
* Added show closed port on zucca
* Fixed -p problem on filippo
* Added ability to get closed port on singsing engine

Version 0.4:
------------------------
* Added filippo.c (socks scanner) on example session
* Added telnet_grabber.c on example session

Version 0.3:
------------------------
* Added SINGSING_SEGMENT_SCAN mode
* Fixed some unckecked variables at start

Version 0.2:
------------------------
* Added SINGSING_NODUP_SCAN scanning mode, remove duplicate results
* Added auto adjust sleep time on send syn procedures
* Fixed a lot of inital bugs
* Added examples section (zucca.c and ssh_scan.c)

Version 0.1:
------------------------
* First release of this project


src/singsing.h
/*****************************************************************************
 * singsing.h is a part of singsing project                                  *
 *                                                                           *
 * $Id::                                                                  $: *
 *                                                                           *
 * Copyright (c) 2007, Agazzini Maurizio - inode@mediaservice.net            *
 * All rights reserved.                                                      *
 *                                                                           *
 * Redistribution and use in source and binary forms, with or without        *
 * modification, are permitted provided that the following conditions        *
 * are met:                                                                  *
 *     * Redistributions of source code must retain the above copyright      *
 *       notice, this list of conditions and the following disclaimer.       *
 *     * Redistributions in binary form must reproduce the above copyright   *
 *       notice, this list of conditions and the following disclaimer in     *
 *       the documentation and/or other materials provided with the          *
 *       distribution.                                                       *
 *     * Neither the name of @ Mediaservice.net nor the names of its         *
 *       contributors may be used to endorse or promote products derived     *
 *       from this software without specific prior written permission.       *
 *                                                                           *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS       *
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT         *
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR     *
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT      *
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,     *
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED  *
 * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR    *
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF    *
 * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING      *
 * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS        *
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.              *
 *****************************************************************************/


#ifndef _SINGSING_H

#define _SINGSING_H

#define SINGSING_VERSION "0.5"

#define SINGSING_REVISION       "$Rev$"

#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <time.h>
#include <string.h>
#include <netdb.h>
#include <pthread.h>
#include <pcap.h>
#include <net/if.h>
#include <sys/ioctl.h>
#include <sys/socket.h>
#include <arpa/inet.h>
#include <fcntl.h>
#include <sys/types.h>
#include <sys/socket.h> 

#ifdef SOLARIS
#include <sys/sockio.h>
#include <arpa/nameser_compat.h>
#endif

#ifdef MAC
#include <netinet/in.h>
#endif

#define SINGSING_BASIC_SCAN 	0
#define SINGSING_NODUP_SCAN	2
#define SINGSING_SEGMENT_SCAN	4
#define SINGSING_SHOW_CLOSED	8

#define SINGSING_TIMEOUT 	30

#define SINGSING_OPEN		0
#define SINGSING_CLOSE		1

struct singsing_result_queue {
	unsigned long ip;
	unsigned int port;
	unsigned char type;
	time_t rec_time;	
	struct singsing_result_queue * next;
};

struct singsing_status_struct {
	unsigned long total_port;
	unsigned long current_port;
	unsigned long synps;
	time_t init_time;
	time_t current_time;
};

struct singsing_descriptor {
	int singsing_band;
	char * singsing_device;
	unsigned long singsing_start_ip;
	unsigned long singsing_end_ip;
	unsigned int singsing_min_port;
	unsigned int singsing_max_port;
	int singsing_socket[500];
	pcap_t * singsing_descr;
	u_short singsing_ipid;
	int singsing_finished;
	unsigned int singsing_ports;
	unsigned long singsing_sleep_band;
	unsigned long singsing_synps;
	int singsing_raw_socket;
	unsigned long singsing_source_ip;
	pthread_t singsing_thread_id[3];
	unsigned long singsing_cur_port;
	unsigned int singsing_scan_mode;

	// Data lists
	struct singsing_port_list * singsing_first_port;
	struct singsing_port_list * singsing_last_port;

	struct singsing_result_queue * singsing_first_result;
	struct singsing_result_queue * singsing_last_result;
	pthread_mutex_t singsing_result_queue_lock ;

	pthread_mutex_t packet_queue_lock;

	struct singsing_packet_queue * singsing_first_packet;
	struct singsing_packet_queue * singsing_last_packet;

	struct singsing_status_struct singsing_cur_status;
};

// Prototypes
int singsing_add_port( struct singsing_descriptor * fd, unsigned int port );
int singsing_set_scan_interface( struct singsing_descriptor * fd, char * interface );
int singsing_set_scan_host( struct singsing_descriptor * fd, char * host);
int singsing_init( struct singsing_descriptor * fd );
void singsing_destroy( struct singsing_descriptor * fd );
void singsing_create( struct singsing_descriptor * fd );
int singsing_scanisfinished( struct singsing_descriptor * fd );
void singsing_set_bandwidth( struct singsing_descriptor * fd, int a);
void singsing_get_status( struct singsing_descriptor * fd, struct singsing_status_struct * cur );
void singsing_set_scanmode( struct singsing_descriptor * fd, int a );
struct singsing_result_queue * singsing_get_result( struct singsing_descriptor * fd );

#endif 



src/singsing_p.h
/*****************************************************************************
 * singsing_p.h is a part of singsing project                                *
 *                                                                           *
 * $Id::                                                                  $: *
 *                                                                           *
 * Copyright (c) 2007, Agazzini Maurizio - inode@mediaservice.net            *
 * All rights reserved.                                                      *
 *                                                                           *
 * Redistribution and use in source and binary forms, with or without        *
 * modification, are permitted provided that the following conditions        *
 * are met:                                                                  *
 *     * Redistributions of source code must retain the above copyright      *
 *       notice, this list of conditions and the following disclaimer.       *
 *     * Redistributions in binary form must reproduce the above copyright   *
 *       notice, this list of conditions and the following disclaimer in     *
 *       the documentation and/or other materials provided with the          *
 *       distribution.                                                       *
 *     * Neither the name of @ Mediaservice.net nor the names of its         *
 *       contributors may be used to endorse or promote products derived     *
 *       from this software without specific prior written permission.       *
 *                                                                           *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS       *
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT         *
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR     *
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT      *
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,     *
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED  *
 * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR    *
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF    *
 * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING      *
 * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS        *
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.              *
 *****************************************************************************/


#ifndef _SINGSING_P_H

#define _SINGSING_P_H

//#define DEBUG 1
// Prototypes
int singsing_checksum(unsigned short* data, int length);
unsigned short singsing_in_cksum_tcp(int src, int dst, \
	unsigned short *addr, int len);
void singsing_packet_rec(u_char *args,const struct pcap_pkthdr* pkthdr, \
	const u_char* packet);
void * singsing_processing_thread(void *parm);
void * singsing_send_syn_thread(void *parm);
int singsing_send_syn( int sock, long dest_ip , long source_ip, long port, struct singsing_descriptor * fd);
void * singsing_sniff_thread(void *parm);
void singsing_set_thread_priority( int priority );
unsigned long singsing_get_ip(struct singsing_descriptor * fd,char* interface) ;
int singsing_bind_port( struct singsing_descriptor * fd,unsigned long ip );



struct singsing_port_list {
	long port;
	struct singsing_port_list * next;	
};

struct singsing_packet_queue {
	const u_char* packet;
	bpf_u_int32 len;
	struct singsing_packet_queue * next;
};


// Global variables


// Fixing solaris
#ifdef SOLARIS
typedef uint8_t u_int8_t;
typedef uint16_t u_int16_t;
typedef uint32_t u_int32_t;
#endif

/* TCP Header structure, taken from linux includes */

struct tcphdr
  {
    u_int16_t th_sport;         /* source port */
    u_int16_t th_dport;         /* destination port */
    u_int32_t th_seq;             /* sequence number */
    u_int32_t th_ack;             /* acknowledgement number */
#  if BYTE_ORDER == LITTLE_ENDIAN
    u_int8_t th_x2:4;           /* (unused) */
    u_int8_t th_off:4;          /* data offset */
#  endif
#  if BYTE_ORDER == BIG_ENDIAN
    u_int8_t th_off:4;          /* data offset */
    u_int8_t th_x2:4;           /* (unused) */
#  endif
    u_int8_t th_flags;
#  define TH_FIN        0x01
#  define TH_SYN        0x02
#  define TH_RST        0x04
#  define TH_PUSH       0x08
#  define TH_ACK        0x10
#  define TH_URG        0x20
    u_int16_t th_win;           /* window */
    u_int16_t th_sum;           /* checksum */
    u_int16_t th_urp;           /* urgent pointer */

    //MSS FIX!
    u_int32_t mss;
};

/* IP Header structure, taken from linux includes */

struct ip
  {
#if BYTE_ORDER == LITTLE_ENDIAN
    unsigned int ip_hl:4;               /* header length */
    unsigned int ip_v:4;                /* version */
#endif
#if BYTE_ORDER == BIG_ENDIAN
    unsigned int ip_v:4;                /* version */
    unsigned int ip_hl:4;               /* header length */
#endif
    u_int8_t ip_tos;                    /* type of service */
    u_short ip_len;                     /* total length */
    u_short ip_id;                      /* identification */
    u_short ip_off;                     /* fragment offset field */
#define IP_RF 0x8000                    /* reserved fragment flag */
#define IP_DF 0x4000                    /* dont fragment flag */
#define IP_MF 0x2000                    /* more fragments flag */
#define IP_OFFMASK 0x1fff               /* mask for fragmenting bits */
    u_int8_t ip_ttl;                    /* time to live */
    u_int8_t ip_p;                      /* protocol */
    u_short ip_sum;                     /* checksum */
    struct in_addr ip_src, ip_dst;      /* source and dest address */
  };

#define ETH_ALEN       6               /* Octets in one ethernet addr   */
#define  ETHERTYPE_IP            0x0800          /* IP */

/* 10Mb/s ethernet header */
struct ether_header
{
  u_int8_t  ether_dhost[ETH_ALEN];      /* destination eth addr */
  u_int8_t  ether_shost[ETH_ALEN];      /* source ether addr    */
  u_int16_t ether_type;                 /* packet type ID field */
};


struct singsing_psd_tcp {
        struct in_addr src;
        struct in_addr dst;
        unsigned char pad;
        unsigned char proto;
        unsigned short tcp_len;
        struct tcphdr tcp;
};

#endif 



src/examples/zucca.c
/*****************************************************************************
 * zucca.c is a part of singsing project                                     *
 *                                                                           *
 * $Id::                                                                  $: *
 *                                                                           *
 * Copyright (c) 2007, Agazzini Maurizio - inode@mediaservice.net            *
 * All rights reserved.                                                      *
 *                                                                           *
 * Redistribution and use in source and binary forms, with or without        *
 * modification, are permitted provided that the following conditions        *
 * are met:                                                                  *
 *     * Redistributions of source code must retain the above copyright      *
 *       notice, this list of conditions and the following disclaimer.       *
 *     * Redistributions in binary form must reproduce the above copyright   *
 *       notice, this list of conditions and the following disclaimer in     *
 *       the documentation and/or other materials provided with the          *
 *       distribution.                                                       *
 *     * Neither the name of @ Mediaservice.net nor the names of its         *
 *       contributors may be used to endorse or promote products derived     *
 *       from this software without specific prior written permission.       *
 *                                                                           *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS       *
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT         *
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR     *
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT      *
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,     *
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED  *
 * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR    *
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF    *
 * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING      *
 * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS        *
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.              *
 *****************************************************************************/

#include "singsing.h"

#include <term.h>

#define VERSION "0.4"

/* global variables */
int host_ports	= 0;
int show_closed = 0;

extern int errno;
struct termios savetty;


/* Prototypes */
int parse_system_service( struct singsing_descriptor * fd );
void parse_port( struct singsing_descriptor * fd, char * ports );
void usage( char * argv0 );
void tty_raw( void );
void tty_normal( void );
int getch( void );

int main( int argc, char ** argv )
{

	char opt;
 	char * ports = NULL;
	char buf[200];

	struct singsing_result_queue * cur_res;
	struct singsing_status_struct current_status;
	struct in_addr result;

	struct singsing_descriptor fd;

	time_t endtime;
	time_t cur;

	struct tm ts;

	singsing_create(&fd);

	while((opt = getopt(argc, argv, "i:b:p:h:ct:s:")) != -1)
	{
		switch (opt)
                {
			case 'i':
				singsing_set_scan_interface( &fd, optarg );
				break;
			case 'b':
				singsing_set_bandwidth( &fd, atoi( optarg ) );
				break;
			case 'h':
				singsing_set_scan_host( &fd, optarg );
				break;
			case 'p':
				ports = optarg;
				break;
			case 'c':
				singsing_set_scanmode( &fd, SINGSING_SHOW_CLOSED );
				break;
			case 's':
// 				sleeper = atoi( optarg );
 				fprintf(stderr, "\noption -s not implemented yet\n");
				break;
			default :
				usage(argv[0]);
				break;
		}
	}

	fprintf(stderr, "\n zucca syn scanner v%s with singsing v%s revision %s\n", VERSION, SINGSING_VERSION, SINGSING_REVISION);
        fprintf(stderr, " by inode@mediaservice.net\n\n");

	host_ports = 0;

	if( ports != NULL ) 
		parse_port( &fd, ports );
	else
		parse_system_service( &fd );

	fprintf(stderr, " Ports per host  : %u\n", host_ports);


	singsing_set_scanmode( &fd, SINGSING_NODUP_SCAN );
	singsing_set_scanmode( &fd, SINGSING_SEGMENT_SCAN );
	 

	fprintf( stderr, " Starting scan...\n\n");

	if( singsing_init(&fd) < 0 ) {
		usage(argv[0]);
	}

	tty_raw();

        do {
                cur_res = singsing_get_result(&fd);
                if( cur_res != NULL ) {
                        result.s_addr = ntohl(cur_res->ip);

			if( cur_res->type == SINGSING_OPEN )
                        	printf(" zucca open %s:%u\n",inet_ntoa( result ), cur_res->port );
			else
                        	printf(" zucca close %s:%u\n",inet_ntoa( result ), cur_res->port );

                        free(cur_res);
                } else {
                        usleep(300000);

			if( getch() > 0 ) {
				singsing_get_status(&fd, &current_status);

				cur = current_status.init_time;
				cur += difftime(current_status.current_time, current_status.init_time) * current_status.total_port / current_status.current_port;
			
				ts = *localtime(&cur);
				
				strftime(buf, sizeof(buf), "%a %Y-%m-%d %H:%M:%S %Z", &ts);
				
				fprintf(stderr," stats: %lu%% expected end at %s\n", \
				100 * current_status.current_port / current_status.total_port, \
				buf);

			}
		}

        } while( singsing_scanisfinished(&fd) != 2 || cur_res != NULL);

	endtime = time( NULL );
	
	singsing_get_status(&fd, &current_status);

	fprintf( stderr, " %lu ports scanned in %.0lf seconds\n", current_status.total_port, difftime(endtime, current_status.init_time));

	tty_normal();

	singsing_destroy(&fd);	

	exit( EXIT_SUCCESS );
}


/* Parse /etc/services */
int parse_system_service( struct singsing_descriptor * fd )
{
	struct servent * tmp;

	while( ( tmp = getservent()) != NULL ) {
		if( strcmp( tmp->s_proto, "tcp") == 0 ) {

			singsing_add_port( fd, ntohs(tmp->s_port) );

			host_ports++;
		}
	}

	endservent();

	return 0;
} 

/* Parse port command line */
void parse_port( struct singsing_descriptor * fd, char * ports )
{
	char * tmp;
	long port;

	if( ( tmp = strchr( ports, ',')	) != NULL ) {
		*tmp = 0;
		tmp++;

		parse_port( fd, ports );
		parse_port( fd, tmp );

		return; 
	}

	if( ( tmp = strchr( ports, '-') ) != NULL ) {
		char number[10];
		int i;

		*tmp = 0;
		tmp++;

		for( i = atol( ports ); i <= atol(tmp); i++) {
			snprintf( number, sizeof(number), "%d", i);
			parse_port( fd, number );
		}
		return;
	}

        if( (port = atol(ports)) > 0 ) {
		singsing_add_port( fd, port );
                host_ports++;

                return;
        } 

	return;
}

void usage( char * argv0 )
{
	fprintf( stderr, " Usage:\n");
	fprintf( stderr, "  %s -h <arg> -i <arg> [-b <arg>] [-p <arg>] [-c]\n\n",argv0);
	fprintf( stderr, " -h Host/s to scan  (ex 192.168.0.0/24)\n");
	fprintf( stderr, " -i Interface\n");
	fprintf( stderr, " -b Usable bandwidth in KB (Default 15)\n");
	fprintf( stderr, " -p Ports (ex 22,23,40-50,99)\n");   
 	fprintf( stderr, " -c Display closed ports\n\n");

	exit( EXIT_FAILURE );
}

void tty_normal(void)
{
        tcsetattr(0, TCSADRAIN, &savetty);
}

void tty_raw(void)
{
	struct termios tty;
        tcgetattr(0, &savetty);
        tcgetattr(0, &tty);
        tty.c_lflag &= ~(ECHO|ECHONL|ICANON|IEXTEN);
        tty.c_cc[VTIME] = 0;
        tty.c_cc[VMIN] = 0;
        tcsetattr(0, TCSADRAIN, &tty);
}

int getch (void)
{
	char buf[2];
        if (read (0, buf, 1)) {
                return buf[0];
        }
        return -1;
}



src/examples/filippo.c
/*****************************************************************************
 * filippo.c is a part of singsing project                                   *
 *                                                                           *
 * $Id::                                                                  $: *
 *                                                                           *
 * Copyright (c) 2007, Agazzini Maurizio - inode@mediaservice.net            *
 * All rights reserved.                                                      *
 *                                                                           *
 * Redistribution and use in source and binary forms, with or without        *
 * modification, are permitted provided that the following conditions        *
 * are met:                                                                  *
 *     * Redistributions of source code must retain the above copyright      *
 *       notice, this list of conditions and the following disclaimer.       *
 *     * Redistributions in binary form must reproduce the above copyright   *
 *       notice, this list of conditions and the following disclaimer in     *
 *       the documentation and/or other materials provided with the          *
 *       distribution.                                                       *
 *     * Neither the name of @ Mediaservice.net nor the names of its         *
 *       contributors may be used to endorse or promote products derived     *
 *       from this software without specific prior written permission.       *
 *                                                                           *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS       *
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT         *
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR     *
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT      *
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,     *
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED  *
 * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR    *
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF    *
 * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING      *
 * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS        *
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.              *
 *****************************************************************************/


#include "singsing.h"

#define NAME "socks"
#define VERSION "0.1"

#define SOCKS_PORT	1080
#define READ_TIMEOUT	30
#define CONNECT_TIMEOUT 30

void usage( char * argv );
int socks_v4_scan(char * host, unsigned int port); 
int socks_v5_scan(char * host, unsigned int port);

unsigned int port = SOCKS_PORT;

int connect_ip[4];
unsigned int target_port;
unsigned int target_port1;


int main(int argc, char ** argv)
{
	char * target = NULL;
	char * device = NULL;
	char * ip = NULL;
	char * star = NULL ;
	char opt;
	struct in_addr result;
        time_t start_time;
        time_t end_time;
	int band = 5;

	struct singsing_result_queue * cur_res;

        struct singsing_descriptor fd;

        singsing_create(&fd);

	while((opt = getopt(argc, argv, "i:h:b:c:p:")) != -1)
	{
		switch (opt)
		{
			case 'i':
				device = optarg;
				break;
			case 'h':
				target = optarg;
				break;
			case 'b':
				band = atoi( optarg );
				break;
			case 'c':
				ip = optarg;
				break;
			case 'p':
				port = atoi( optarg );
				break;
			default:
				usage( argv[0] );
		}
	}

	if( target == NULL || device == NULL || ip == NULL )
		usage( argv[0] );

	star = strchr( ip, '.');
	*star = 0x0;
	connect_ip[0] = atoi(ip);
	ip = star + 1;

        star = strchr( ip, '.');
        *star = 0x0;
        connect_ip[1] = atoi(ip);
        ip = star + 1;

        star = strchr( ip, '.');
        *star = 0x0;
        connect_ip[2] = atoi(ip);
        ip = star + 1;

        star = strchr( ip, ':');
        *star = 0x0;
        connect_ip[3] = atoi(ip);
        ip = star + 1;

	target_port = atol( ip );

	singsing_set_scan_interface( &fd, device );

	singsing_set_bandwidth( &fd, band );

	singsing_set_scan_host( &fd, target );

	singsing_add_port( &fd, port );

        singsing_set_scanmode( &fd, SINGSING_NODUP_SCAN );

        singsing_set_scanmode( &fd, SINGSING_SEGMENT_SCAN );

	fprintf( stderr, "Starting scan...\n");

	start_time = time(NULL);

	singsing_init(&fd);

	do {
		cur_res = singsing_get_result(&fd);
		if( cur_res != NULL ) {
			result.s_addr = ntohl(cur_res->ip);

			socks_v4_scan(strdup(inet_ntoa( result )), port);
			socks_v5_scan(strdup(inet_ntoa( result )), port);
			
			fflush(stderr);
			fflush(stdout);
			
			free(cur_res);
		} else
                	usleep(300000);
 
	} while( singsing_scanisfinished(&fd) != 2 || cur_res != NULL);


        end_time = time(NULL);

        fprintf( stderr, "\n Scan end in %.0lf seconds\n\n", difftime(end_time, start_time));

        singsing_destroy(&fd);

	return 0;
}


void usage( char * argv )
{
	fprintf(stderr, "\n Usage: %s -i <arg> -h <arg> -c <arg> [-b <arg> -p <arg>]\n", argv);
	fprintf(stderr, "\t-i Interface\n");
	fprintf(stderr, "\t-h Target (CIDR format)\n");
	fprintf(stderr, "\t-b Bandwidth (Default 5KB/s)\n");
	fprintf(stderr, "\t-p Port (Default 1080)\n");
	fprintf(stderr, "\t-c Try connect to (ip:port)\n");
	exit(0);
} 

int net_connect( char * host, int port)
{
	int sd;
	struct sockaddr_in servAddr;
	int flags, flags_old, retval;	
	unsigned int sock_len;
	struct sockaddr_in sin;
	struct timeval tv;
	fd_set rfds;

	servAddr.sin_family = AF_INET;
	//memcpy((char *) &servAddr.sin_addr.s_addr, h->h_addr_list[0], h->h_length);
	servAddr.sin_addr.s_addr = inet_addr(host);
	servAddr.sin_port = htons( port );

	sd = socket(AF_INET, SOCK_STREAM, 0);
	if(sd<0) {
		perror("cannot open socket ");
		exit(1);
	}

	// Set Non Blocking Socket
	flags_old = fcntl( sd, F_GETFL,0);
	flags = flags_old;
	flags |= O_NONBLOCK;
	fcntl( sd, F_SETFL, flags);

	if( connect(sd, (struct sockaddr *) &servAddr, sizeof(servAddr)) == 0) {
		fcntl( sd, F_SETFL, flags_old);
		return sd;
	}
	
	// Set timeout
	tv.tv_sec = CONNECT_TIMEOUT;
	tv.tv_usec = 0;

	FD_ZERO(&rfds);
	FD_SET(sd, &rfds);
	
	retval = select(FD_SETSIZE, NULL, &rfds, NULL, &tv);

	// if retval < 0 error
	if( retval < 0 ) {
		close( sd );
		return -1;
	}
	sock_len = sizeof( sin );

	// Check if port closed
	if( retval ) {
		if( getpeername( sd, (struct  sockaddr  *) &sin, &sock_len) < 0 ) {
			close( sd );
			return -1;
		} else {
			// XXX
			fcntl( sd, F_SETFL, flags_old);
			return sd;
		}
	}
	close( sd );
	return -1;

} 


int socks_v4_scan(char * host, unsigned int port)
{
	int sock, flags,i,l;
	char buff[2000];
        struct timeval tv;
	char * p;
        time_t cur_time;
        time_t start_time;

	fd_set rfds; 

	memset( buff, 0, sizeof(buff));

	p = buff;
	l = 0;

	sock = net_connect(host, port);

	if( sock < 0 )
		return 0;

	// Testing V4 protocol
/*    * field 1: SOCKS version number, 1 byte, must be 0x04 for this version
    * field 2: command code, 1 byte:
          o 0x01 = establish a TCP/IP stream connection
          o 0x02 = establish a TCP/IP port binding
    * field 3: network byte order port number, 2 bytes
    * field 4: network byte order IP address, 4 bytes
    * field 5: the user ID string, variable length, terminated with a null (0x00)*/
	/*0x04 | 0x01 | 0x00 0x50 | 0x42 0x66 0x07 0x63 | 0x46 0x72 0x65 0x64 0x00*/


	memcpy(buff, "\x04\x01", 2);

	target_port1 = htons( target_port );

	memcpy(buff + 2, &target_port1, 2);
	memcpy(buff + 4, &connect_ip[0], 1);
        memcpy(buff + 5, &connect_ip[1], 1);
        memcpy(buff + 6, &connect_ip[2], 1);
        memcpy(buff + 7, &connect_ip[3], 1);
	memcpy(buff + 8 , "\x46\x72\x65\x64\x00",5);


	write(sock,buff,13);

        tv.tv_sec = READ_TIMEOUT;
        tv.tv_usec = 0;

        flags = fcntl( sock, F_GETFL,0);
        flags |= O_NONBLOCK;
        fcntl( sock, F_SETFL, flags);

	FD_ZERO( &rfds );
	FD_SET( sock, &rfds );

        start_time = time(NULL);

	while( select( FD_SETSIZE , &rfds, (fd_set *) 0, (fd_set *) 0, &tv) > 0) {
		i = read( sock, buff ,sizeof( buff )); 
		l += i;

		cur_time = time(NULL);

		if( i < 0) {
			close( sock );
			return 0;
		}

		if( l >= 8) 
			break;
			
		// resolving CLOSE_WAIT problems
		if( difftime(cur_time, start_time) > READ_TIMEOUT) {
			close( sock );
			return 0;
		}

		
			
		usleep(3000);
		p += i;
	}

	if( memcmp(buff,"\x00\x5a", 2 ) == 0 ) {
		fprintf(stderr,"Socks v4: %s (WORKING)\n", host);
		close(sock);

		return 1;
	}

	if( memcmp(buff,"\x00\x5c", 2 ) == 0 ) 
		fprintf(stderr,"Socks v4: %s (REQ IDENTD)\n", host);

	if( memcmp(buff,"\x00\x5d", 2 ) == 0 )
		fprintf(stderr,"Socks v4: %s (IDENTD USER)\n", host);

	close(sock);
	return 0;
} 


int socks_v5_scan(char * host, unsigned int port)
{
        int sock, flags,i,l;
        char buff[2000];
        struct timeval tv;
        char * p;
        time_t cur_time;
        time_t start_time;

        fd_set rfds;

        memset( buff, 0, sizeof(buff));

        p = buff;
        l = 0;

        sock = net_connect(host, port);

        if( sock < 0 )
                return 0;

        // Testing V5 protocol
	// Requesting socks noauth.

        write(sock,"\x05\x01\x00",3);

        tv.tv_sec = READ_TIMEOUT;
        tv.tv_usec = 0;

        flags = fcntl( sock, F_GETFL,0);
        flags |= O_NONBLOCK;
        fcntl( sock, F_SETFL, flags);

        FD_ZERO( &rfds );
        FD_SET( sock, &rfds );

        start_time = time(NULL);

        while( select( FD_SETSIZE , &rfds, (fd_set *) 0, (fd_set *) 0, &tv) > 0) {
                i = read( sock, buff ,sizeof( buff ));
                l += i;

                cur_time = time(NULL);

                if( i < 0) {
                        close( sock );
                        return 0;
                }

                if( l >= 2)
                        break;

                // resolving CLOSE_WAIT problems
                if( difftime(cur_time, start_time) > READ_TIMEOUT) {
                        close( sock );
                        return 0;
                }



                usleep(3000);
                p += i;
        }

        if( memcmp(buff,"\x05\xFF", 2 ) == 0 ) {
                close(sock);

                return 0;
        }

        memcpy(buff, "\x05\x01\x00\x01", 4);

        target_port1 = ntohs( target_port );

        memcpy(buff + 4, &connect_ip[0], 1);
        memcpy(buff + 5, &connect_ip[1], 1);
        memcpy(buff + 6, &connect_ip[2], 1);
        memcpy(buff + 7, &connect_ip[3], 1);
        memcpy(buff + 8, &target_port1, 2);

        write(sock,buff,10);

        tv.tv_sec = READ_TIMEOUT;
        tv.tv_usec = 0;

        flags = fcntl( sock, F_GETFL,0);
        flags |= O_NONBLOCK;
        fcntl( sock, F_SETFL, flags);

        FD_ZERO( &rfds );
        FD_SET( sock, &rfds );

        start_time = time(NULL);

        while( select( FD_SETSIZE , &rfds, (fd_set *) 0, (fd_set *) 0, &tv) > 0) {
                i = read( sock, buff ,sizeof( buff ));
                l += i;

                cur_time = time(NULL);

                if( i < 0) {
                        close( sock );
                        return 0;
                }

                if( l >= 10)
                        break;

                // resolving CLOSE_WAIT problems
                if( difftime(cur_time, start_time) > READ_TIMEOUT) {
                        close( sock );
                        return 0;
                }



                usleep(3000);
                p += i;
        }

        if( memcmp(buff,"\x05\x00", 2 ) == 0 ) 
                fprintf(stderr,"Socks v5: %s (WORKING)\n", host);

        close(sock);
        return 1;
}
